### PR TITLE
3864 transferrable vaults

### LINKED ISSUE #3864: make Treasury Vaults transferrable
## What is the Problem Being Solved?

RUN Protocol Vaults operate as an ongoing contract, where the owner of the vault holds a seat in the ongoing activity of the vault. Our general model of smart contracts is that such "Seats" should be transferable. We should accomplish that with our initial smart contract use case.

Vaults contain collateral for a "loan". Loans start out over-collateralized, so the vaults can represent a valuable asset. The asset should be transferrable using standard ERTP approaches. If buyers believe that the value of the collateral will grow faster than the required interest, it could make sense to purchase it.

## Description of the Design

The owner of the vault should have the ability to request a transfer. This would prevent further withdrawals, and return an invitation with sufficient specificity that a potential buyer could evaluate the contents.

The basic idea is that without further support from Zoe, the contract can create an invitation that when completed would turn all the assets (stored on the internal `vaultSeat`) over to a new seat. The current assertVaultIsOpen() call would be updated to also check a generation number if the idea is to transfer the vault object to the new owner rather than reallocating all the assets.

## Security Considerations

Standard ERTP visibility issues. There doesn't seem to be financial risk to the system. The owner can't extract collateral unless the loan is overcollateralized. The new owner doesn't have more access to the contents than the original owner did.

## Test Plan

nothing special.

### PR DESCRIPTION
**Pull request by dtribble**
closes: #3864
refs: #4341

## Description

This builds on top of #4341.

Implementation of transferrable vaults. This refactors vaults into an `InnerVault` and a wrapping  `OuterVault`, where the inner vault has the authority and the outer vault proxies operations and can be revoked. The outer is revoked when a transfer is started, and MUST stop multi-turn operations since the transfer invitation should capture the state of the vault as of the point of transfer.

Key design points
* the `outerVault` pointer gets nulled on revocation to enforce revocation by eliminating reachability
* creation of the  `vaultKit` is moved into `vault.js`
* the `prioritizedVault` collection just has pointers to the `innerVault`
* the outer vault is not reachable from `innerVault`, only upon creation
* The UI notifier from before `transfer` gets `finished` with the `transfer` state and a new one is opened for the new outerVault.

### Security Considerations

The simplification should make it more straightforward to ensure that vaults are isolated. 

All operations that have `await` before state change MUST check for an interleaving `transfer`. The only one right now is `adjustBalances`.

### Documentation Considerations

* Need to add `makeTransferInvitation` to the Vault documentation on our site.
* It would be good to describe the transfer architecture generally

### Testing Considerations

* prior vault is revoked after initiating transfer
* transfer invitation has vault status details [not done yet]
* prior notifier is `finished` with new `TRANSFER` state
* new notifier is `ACTIVE` state
* payout of RUN from original seat is still available
* a `transfer` started while an `adjust` is in progress will abort the adjust (interleaving test)

### CODE REVIEW

File: packages/run-protocol/src/vaultFactory/prioritizedVaults.js [hunk 1]
```diff
@@ -48,7 +49,7 @@ export const makePrioritizedVaults = reschedulePriceCheck => {
   // To deal with fluctuating prices and varying collateralization, we schedule a
   // new request to the priceAuthority when some vault's debtToCollateral ratio
   // surpasses the current high-water mark. When the request that is at the
-  // current high-water mark fires, we reschedule at the new highest ratio
+  // current high-water mark fires, we reschedule
   // (which should be lower, as we will have liquidated any that were at least
   // as high.)
```
Comment (Chris-Hibbert) [inline packages/run-protocol/src/vaultFactory/prioritizedVaults.js]:
```suggestion
  // current high-water mark fires, we reschedule at the new (presumably
  // lower) rate.
```

File: packages/run-protocol/src/vaultFactory/vault.js [hunk 2]
```diff
@@ -572,14 +617,30 @@ export const makeVaultKit = (
...
+  const setupOuter = inner => {
+    const { vault, uiUpdater: updater } = makeOuterKit(inner);
+    outerUpdater = updater;
+    updateUiState();
+    return harden({
+      uiNotifier: vault.getNotifier(),
+      invitationMakers: Far('invitation makers', {
+        AdjustBalances: vault.makeAdjustBalancesInvitation,
+        CloseVault: vault.makeCloseInvitation,
+        // TransferVault: vault.makeTransferVaultInvitation,
```
Comment (Chris-Hibbert) [inline packages/run-protocol/src/vaultFactory/vault.js]: why commented out?
```suggestion
        // TransferVault: vault.makeTransferInvitation,
```

File: packages/run-protocol/src/vaultFactory/vaultManager.js [hunk 3]
```diff
@@ -365,34 +365,24 @@ export const makeVaultManager = (
...
       runMint,
       priceAuthority,
     );
-    const {
-      vault,
-      actions: { openLoan },
-    } = vaultKit;
+
+    // Don't record the vault until it gets opened
+    // TODO
```
Comment (Chris-Hibbert) [inline packages/run-protocol/src/vaultFactory/vaultManager.js]:
```suggestion
    // TODO: Don't record the vault until it gets opened
```

File: packages/run-protocol/test/vaultFactory/test-vaultFactory.js [hunk 4]
```diff
@@ -1245,6 +1249,142 @@ test('adjust balances', async t => {
...
+  t.deepEqual(collateralAmount, collateralAfter2, 'vault has 1000n aEth');
+
+  const transferFinish = await E(transferNotifier).getUpdateSince();
+  t.deepEqual(
+    transferFinish.value.vaultState,
+    VaultState.TRANSFER,
+    't2 closed old notifier',
+  );
+
+  const t2Status = await E(t2Notifier).getUpdateSince();
```
Comment (Chris-Hibbert) [inline packages/run-protocol/test/vaultFactory/test-vaultFactory.js:1375]: show that `getDebtAmount()`, for instance, fails on the old vault?

**Review by Chris-Hibbert — APPROVED** [review]
Looks good to me.

File: packages/run-protocol/src/vaultFactory/prioritizedVaults.js (hunk 1 above)
Comment (dtribble) [reply packages/run-protocol/src/vaultFactory/prioritizedVaults.js]: done

File: packages/run-protocol/src/vaultFactory/vault.js (hunk 2 above)
Comment (dtribble) [reply packages/run-protocol/src/vaultFactory/vault.js]: there was a type issue. Fixed.

File: packages/run-protocol/src/vaultFactory/vaultManager.js (hunk 3 above)
Comment (dtribble) [reply packages/run-protocol/src/vaultFactory/vaultManager.js]: fixed

File: packages/run-protocol/test/vaultFactory/test-vaultFactory.js (hunk 4 above)
Comment (dtribble) [reply packages/run-protocol/test/vaultFactory/test-vaultFactory.js:1375]: Already tested for the first transfer on line 1351

Comment (dtribble) [reply packages/run-protocol/test/vaultFactory/test-vaultFactory.js:1375]: added anyway

File: packages/run-protocol/src/vaultFactory/vaultManager.js (hunk 3 above)
Comment (dtribble) [reply packages/run-protocol/src/vaultFactory/vaultManager.js:378]: fixed

File: packages/run-protocol/src/vaultFactory/vault.js (hunk 2 above)
Comment (dtribble) [reply packages/run-protocol/src/vaultFactory/vault.js:629]: fixed

File: packages/run-protocol/src/vaultFactory/prioritizedVaults.js (hunk 1 above)
Comment (Chris-Hibbert) [reply packages/run-protocol/src/vaultFactory/prioritizedVaults.js]: Change is not visible here.

File: packages/run-protocol/src/vaultFactory/vaultFactory.js [hunk 5]
```diff
@@ -242,7 +242,8 @@ export const start = async (zcf, privateArgs) => {
 
   /** @type {VaultFactoryPublicFacet} */
   const publicFacet = Far('vaultFactory public facet', {
-    makeLoanInvitation,
+    makeLoanInvitation: makeVaultInvitation, // deprecated
```
Comment (turadg) [inline packages/run-protocol/src/vaultFactory/vaultFactory.js:245]: when can we remove it?

File: packages/run-protocol/src/vaultFactory/vault.js [hunk 6]
```diff
@@ -31,14 +31,43 @@ const trace = makeTracer('Vault');
...
   LIQUIDATING: 'liquidating',
   CLOSED: 'closed',
+  TRANSFER: 'transfer',
+};
+
+const makeOuterKit = inner => {
+  const { updater: uiUpdater, notifier } = makeNotifierKit();
+
+  const assertActive = v => {
+    // console.log('OUTER', v, 'INNER', inner);
```
Comment (turadg) [inline packages/run-protocol/src/vaultFactory/vault.js:49]: no disabled code

**Review by turadg — APPROVED** [review]
Definite improvement.

I think we should try to come up with more explicit language than "inner" vs "outer" vaults but that's easy to change in the next refactoring.

**Pull request merged**